hw012/update_mailroom.py

```diff
@@ -0,0 +1,147 @@
+donor_list = {
+    'Margie DeBella': [100, 95],
+    'Liz Rogers': [1000, 5000, 25],
+    'Lisa Stokes': [800],
+    'Fred Flinstone': [2000, 400, 1500]
+}
+
+
+def start_screen():
+    """Opening screen to inform user of program options."""
+    opener = "Welcome to Mailroom Madness"
+    menu_option = """
+    Choose from the following:\n
+    T - Send a (T)hank You
+    R - Create a (R)eport
+    quit - Quit the program\n
+    Input:  """
+
+    print(opener)
+
+    path = input(menu_option)
+    check_valid_path(path)
+
+    while path:
+        path = input("Please enter from the selections below:\n" + menu_option)
+        path = check_valid_path(path)
+
+
+def check_valid_path(path):
+    unvalid = True
+    try:
+        path = path.lower()
+        if path == "q" or path == "quit":
+            quit()
+        elif path == "r":
+            generate_report()
+        elif path == "t":
+            thankyou()
+        else:
+            return unvalid
+    except TypeError:
+        return unvalid
+
+
+def thankyou():
+    menu_option = """
+    Please enter a name or choose from the following:\n
+    list - Print a list of previous donors
+    quit - Return to the main menu\n
+    Input: """
+
+    print(menu_option)
+
+    donor = input()
+    check_thankyou_path(donor)
+
+
+def check_thankyou_path(donor):
+
+    if donor.lower() == "quit":
+        start_screen()
+    elif donor.lower() == "list":
+        print("See below for a list of donor names")
+        print(donor_name_only())
+        thankyou()
+    try:
+        if donor_list[donor]:
+            donor_add(donor)
+    except KeyError:
+        donor_add(donor)
+
+
+def generate_report():
+    """Generates a report for user with key information
+       still doesn't print sorted"""
+    print("Name \t\t|Total \t|# | Average\n" + ("_" * 40))
+
+    sorted_list = sorted(donor_list, key=lambda donor: sum(donor_list[donor]),
+                         reverse=True)
+    for key in sorted_list:
+        donor_name = key
+        total_donated = sum(donor_list[key])
+        num_donate = len(donor_list[key])
+        avg = total_donated / num_donate
+        print('%s \t|$%d \t|%d | $%d' %
+              (donor_name, total_donated, num_donate, avg))
+    return False
+    start_screen()
+
+
+def donor_name_only():
+    """Prints list of donor names from dictionary."""
+    for key in donor_list:
+        print(key)
+
+
+def donor_add(donor):
+    """Add donor to database"""
+    if donor not in donor_list:
+        donor_list[donor] = []
+        donation = donation_add(donor)
+        while not isinstance(donation, float) or donation < 0:
+            donation = donation_add(donor)
+
+    donor_list[donor].append[donation]
```

Comment:
Attached is a screenshot of sublime.

It runs in python3 in the terminal, but when I get to adding a donor name i
get the following:

[hw012 *]

Samanthas-MBP:hw012 samanthadebella$ python3 update_mailroom.py

Welcome to Mailroom Madness


    Choose from the following:


    T - Send a (T)hank You

    R - Create a (R)eport

    quit - Quit the program


    Input:  t


    Please enter a name or choose from the following:


    list - Print a list of previous donors

    quit - Return to the main menu


    Input:

andrew prince

Traceback (most recent call last):

  File "update_mailroom.py", line 67, in check_thankyou_path

    if donor_list[donor]:

KeyError: 'andrew prince'


During handling of the above exception, another exception occurred:


Traceback (most recent call last):

  File "update_mailroom.py", line 147, in <module>

    start_screen()

  File "update_mailroom.py", line 22, in start_screen

    check_valid_path(path)

  File "update_mailroom.py", line 38, in check_valid_path

    thankyou()

  File "update_mailroom.py", line 55, in thankyou

    check_thankyou_path(donor)

  File "update_mailroom.py", line 70, in check_thankyou_path

    donor_add(donor)

  File "update_mailroom.py", line 100, in donor_add

    donor_list[donor].append(donation)

KeyError: 'andrew prince'

Then I tried: def donor_add(donor, donation) and when I run it in python3 i
still cant get it to print a letter to the terminal.

On Wed, Jul 15, 2015 at 12:52 PM, Joel Stanner <notifications@github.com>
wrote:

> In hw012/update_mailroom.py
> <https://github.com/samanthaprince/sea-c42-python/pull/8#discussion_r34721271>
> :
>
> > +
> > +def donor_name_only():
> > +    """Prints list of donor names from dictionary."""
> > +    for key in donor_list:
> > +        print(key)
> > +
> > +
> > +def donor_add(donor):
> > +    """Add donor to database"""
> > +    if donor not in donor_list:
> > +        donor_list[donor] = []
> > +        donation = donation_add(donor)
> > +        while not isinstance(donation, float) or donation < 0:
> > +            donation = donation_add(donor)
> > +
> > +    donor_list[donor].append[donation]
>
> Can you reproduce the error? It works for me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/samanthaprince/sea-c42-python/pull/8/files#r34721271>.
>